crates/paralegal-policy/src/flows_to.rs

```diff
@@ -166,69 +166,94 @@ impl CtrlFlowsTo {
-                let new_call_site: CallSiteOrDataSink = cs.clone().into();
-                ctrl_flows_to.insert(src_idx, &new_call_site);
-                // initialize with flows from the DataSource to all of the CallSite's DataSinks
-                for sink in callsites_to_callargs
-                    .row_set(&sinks.index(&new_call_site))
-                    .iter()
-                {
-                    ctrl_flows_to.insert(src_idx, sink);
-                }
-            }
-        }
-
         CtrlFlowsTo {
             sources,
             sinks,
             callsites_to_callargs,
             data_flows_to,
-            ctrl_flows_to,
         }
     }
 
     /// Returns whether src->sink is in the transitive closure of data and control flow.
-    pub fn data_and_control_flows_to(&self, src: DataSource, sink: CallSiteOrDataSink) -> bool {
-        let mut queue = vec![src];
-        let mut seen = std::collections::HashSet::<&CallSiteOrDataSink>::new();
+    pub fn data_and_control_flows_to(
+        &self,
+        ctrl: &Ctrl,
+        src: DataSource,
+        sink: CallSiteOrDataSink,
+    ) -> bool {
+        let mut influencees = DataAndControlInfluencees::new(src, ctrl, self);
+        influencees.contains(&sink)
+    }
+}
 
-        while let Some(cur_src) = queue.pop() {
-            if let DataSource::FunctionCall(callsite) = &cur_src {
-                if CallSiteOrDataSink::CallSite(callsite.clone()) == sink {
-                    return true;
-                }
-            }
+pub(crate) struct DataAndControlInfluencees<'a> {
+    to_return: Vec<CallSiteOrDataSink>,
+    queue: Vec<DataSource>,
+    seen: std::collections::HashSet<CallSiteOrDataSink>,
+    ctrl: &'a Ctrl,
+    flows_to: &'a CtrlFlowsTo,
+}
 
-            let cur_src_index = cur_src.clone().to_index(&self.sources);
-            for cur_sink in self
-                .data_flows_to
-                .row_set(&cur_src_index)
-                .iter()
-                .chain(self.ctrl_flows_to.row_set(&cur_src_index).iter())
-            {
-                if &sink == cur_sink {
-                    return true;
-                }
+impl<'a> DataAndControlInfluencees<'a> {
+    pub(crate) fn new(src: DataSource, ctrl: &'a Ctrl, flows_to: &'a CtrlFlowsTo) -> Self {
+        let queue = vec![src];
+        let seen = std::collections::HashSet::<CallSiteOrDataSink>::new();
+
+        DataAndControlInfluencees {
+            to_return: Vec::new(),
+            queue,
+            seen,
+            ctrl,
+            flows_to,
+        }
+    }
+}
+
+impl<'a> Iterator for DataAndControlInfluencees<'a> {
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(r) = self.to_return.pop() {
+            return Some(r);
+        }
+        if let Some(cur_src) = self.queue.pop() {
+            let cur_src_index = cur_src.clone().to_index(&self.flows_to.sources);
+            for cur_sink in self.flows_to.data_flows_to.row_set(&cur_src_index).iter() {
```

Comment:
Can you please add a comment like this so we don't loose sight of this.

```suggestion
            // TODO: We are using a lookup into the index here. We could instead 
            // query the raw SPDG. It is not clear which is more efficient and we 
            // should benchmark this.
            for cur_sink in self.flows_to.data_flows_to.row_set(&cur_src_index).iter() {
```